rust/parsing.rs

```diff
@@ -0,0 +1,945 @@
+                                if duration.weeks != 0 {
+                                    return Err(self.parse_error(
+                                        "Week format durations cannot have days".to_string(),
+                                    ));
+                                }
+
+                                duration.days += value;
+                                if let Some(fraction) = op_fraction {
+                                    let extra_hours = fraction * 24.0;
+                                    let extra_full_hours = extra_hours.trunc();
+                                    duration.hours += extra_full_hours as u32;
+                                    let extra_minutes =
+                                        ((extra_hours - extra_full_hours) * 60.0).round();
+                                    let extra_full_minutes: f64 = extra_minutes.trunc();
+                                    duration.minutes += extra_full_minutes as u32;
+                                    let extra_seconds =
+                                        ((extra_minutes - extra_full_minutes) * 60.0).round();
+                                    let extra_full_seconds = extra_seconds.trunc();
+                                    duration.seconds += extra_full_seconds as u32;
+                                    let micro_extra = ((extra_seconds - extra_full_seconds)
+                                        * 1_000_000.0)
+                                        .round()
+                                        as u32;
+                                    duration.microseconds += micro_extra;
+                                }
+                            }
+                            _ => {
+                                return Err(
+                                    self.parse_error("Invalid duration time unit".to_string())
+                                )
+                            }
+                        }
+                    }
+                }
+            }
+            self.inc();
+
+            if self.end() {
+                break;
+            }
+        }
+
+        parsed.duration = Some(duration);
+
+        Ok(())
+    }
+
+    fn parse_duration_number_frac(&mut self) -> Result<(u32, Option<f64>), ParseError> {
+        let value = self.parse_duration_number()?;
+        if self.current == '.' || self.current == ',' {
+            let mut decimal = 0_f64;
+            let mut denominator = 1_f64;
+            loop {
+                self.inc();
+
+                match self.current {
+                    c if c.is_ascii_digit() => {
+                        decimal *= 10.0;
+                        decimal += f64::from(c.to_digit(10).unwrap());
+                        denominator *= 10.0;
+                    }
+                    _ => return Ok((value, Some(decimal / denominator))),
+                }
+            }
+        } else {
+            Ok((value, None))
+        }
```

Comment:
Refactor the duplicated return value.
Plus same as line 856.
```suggestion
        let fraction = if self.current == '.' || self.current == ',' {
            let mut decimal = 0_f64;
            let mut denominator = 1_f64;

            while let Some(digit) = self.inc().and_then(|ch| ch.to_digit(10)) {
                decimal *= 10.0;
                decimal += f64::from(digit);
                denominator *= 10.0;
            }

            Some(decimal / denominator)
        } else {
            None
        };

        Ok((value, fraction))
```
With the `matches!` change you can get rid of the else case using `.then`, but that's highly opinionated so feel free to reject it.
```suggestion
        let fraction = matches!(self.current, '.' | ',').then(|| {
            let mut decimal = 0_f64;
            let mut denominator = 1_f64;

            while let Some(digit) = self.inc().and_then(|ch| ch.to_digit(10)) {
                decimal *= 10.0;
                decimal += f64::from(digit);
                denominator *= 10.0;
            }

            decimal / denominator
        });

        Ok((value, fraction))
```